src/main/java/com/joejoe2/chat/models/GroupChannel.java

```diff
@@ -96,9 +170,78 @@ public void acceptInvitation(User invitee) throws InvalidOperation {
     lastMessage = joinMessage;
   }
 
+  /**
+   * Let admin(see {@link #checkIsAdmin}) ban target user(may be not in members), cannot {@link
+   * #addMessage} or be {@link #invite invited} until unbanned
+   *
+   * @param admin admin
+   * @param target target user
+   */
+  public void ban(User admin, User target) throws InvalidOperation {
+    checkActionOnSameUser(admin, target);
+    checkIsAdmin(admin, true);
+    if (administrators.contains(target))
+      throw new InvalidOperation("cannot editBanned target because it is an administrator !");
+
+    banned.add(target);
+    GroupMessage banMessage = GroupMessage.banMessage(this, admin, target);
+    messages.add(banMessage);
+    lastMessage = banMessage;
+  }
+
+  /**
+   * Let admin(see {@link #checkIsAdmin}) unban target user(may be not in members)
+   *
+   * @param admin admin
+   * @param target target user
+   */
+  public void unban(User admin, User target) throws InvalidOperation {
+    checkActionOnSameUser(admin, target);
+    checkIsAdmin(admin, true);
+    if (!banned.contains(target)) throw new InvalidOperation("target user has not been banned !");
+
+    banned.remove(target);
+    GroupMessage unbanMessage = GroupMessage.unbanMessage(this, admin, target);
+    messages.add(unbanMessage);
+    lastMessage = unbanMessage;
+  }
+
+  /**
+   * Let admin add target user to administrators, no op if target user is an administrator
+   *
+   * @param admin admin
+   * @param target target user
+   */
+  public void addToAdministrators(User admin, User target) throws InvalidOperation {
+    checkActionOnSameUser(admin, target);
+    if (!administrators.contains(admin))
+      throw new InvalidOperation("admin is not an valid administrator in the channel !");
+    if (!members.contains(target))
+      throw new InvalidOperation("target user is not in members of the channel !");
+    checkIsNotBanned(target);
+
+    administrators.add(target);
+  }
+
+  /**
+   * Let admin remove target user from administrators, no op if target user is not an administrator
+   *
+   * @param admin admin
+   * @param target target user
+   */
+  public void removeFromAdministrators(User admin, User target) throws InvalidOperation {
+    checkActionOnSameUser(admin, target);
+    if (!administrators.contains(admin))
```

Comment:
This is still kinda copy-paste of lines 217 - 218

Reply:
I miss the cases of checkIsAdmin(admin, false)

Reply:
Sure, no worries dude :) It happens ;)